tests/torch/test_torch_trainer.py

```diff
@@ -0,0 +1,123 @@
+        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))
+        x = x.view(-1, 320)
+        x = F.relu(self.fc1(x))
+        x = F.dropout(x, training=self.training)
+        x = self.fc2(x)
+        return F.log_softmax(x, dim=0)
+
+
+@pytest.fixture(scope='module')
+def mnist_datasets():
+    """Parse MNIST datasets."""
+    if not os.environ.get('MNIST_PATH'):
+        logging.warning("MNIST dataset not found locally. I have to download it!")
+
+    dataset_path = os.environ.get('MNIST_PATH', MNIST_PATH)
+    train_set = datasets.MNIST(
+        dataset_path,
+        train=True,
+        download=True,
+        transform=transforms.Compose([
+            transforms.ToTensor(),
+            transforms.Normalize((0.1307,), (0.3081,))
+        ]))
+    val_set = datasets.MNIST(
+        dataset_path,
+        train=False,
+        download=False,
+        transform=transforms.Compose([
+            transforms.ToTensor(),
+            transforms.Normalize((0.1307,), (0.3081,))
+        ]))
+    return train_set, val_set
+
+
+@pytest.mark.hpc
+@pytest.mark.torch_dist
+def test_distributed_trainer_ddp_mnist(mnist_datasets, ddp_strategy):
+    """Test TorchTrainer on MNIST with DDP strategy."""
+    training_config = dict(
+        optimizer='sgd',
+        loss='nllloss'
+    )
+    trainer = TorchTrainer(
+        model=Net(),
+        config=training_config,
+        epochs=2,
+        strategy='ddp',
+        checkpoint_every=1
+    )
+    # Patch strategy with already initialized one to avoid re-initialization
+    trainer.strategy = ddp_strategy
+
+    train_set, val_set = mnist_datasets
+    trainer.execute(train_set, val_set)
+
+
+@pytest.mark.hpc
+@pytest.mark.deepspeed_dist
+def test_distributed_trainer_deepspeed_mnist(mnist_datasets, deepspeed_strategy):
+    """Test TorchTrainer on MNIST with DeepSpeed strategy."""
+    training_config = dict(
+        optimizer='sgd',
+        loss='nllloss'
+    )
+    trainer = TorchTrainer(
+        model=Net(),
+        config=training_config,
+        epochs=2,
+        strategy='deepspeed',
+        checkpoint_every=1
+    )
+    # Patch strategy with already initialized one to avoid re-initialization
+    trainer.strategy = deepspeed_strategy
+
+    train_set, val_set = mnist_datasets
+    trainer.execute(train_set, val_set)
+
+
+@pytest.mark.hpc
+@pytest.mark.horovod_dist
+def test_distributed_trainer_horovod_mnist(mnist_datasets, horovod_strategy):
+    """Test TorchTrainer on MNIST with Horovod strategy."""
+    training_config = dict(
+        optimizer='sgd',
+        loss='nllloss'
+    )
+    trainer = TorchTrainer(
+        model=Net(),
+        config=training_config,
+        epochs=2,
+        strategy='horovod',
+        checkpoint_every=1
+    )
+    # Patch strategy with already initialized one to avoid re-initialization
+    trainer.strategy = horovod_strategy
+
+    train_set, val_set = mnist_datasets
+    trainer.execute(train_set, val_set)
```

Comment:
You could consider parametrising these tests, like this: 
```
@pytest.mark.hpc
@pytest.mark.parametrize(
    'strategy_name,strategy_fixture',
    [
        ('ddp', 'ddp_strategy'),
        ('deepspeed', 'deepspeed_strategy'),
        ('horovod', 'horovod_strategy')
    ]
)
def test_distributed_trainer_mnist(mnist_datasets, request, strategy_name, strategy_fixture):
    """Test TorchTrainer on MNIST with different distributed strategies."""
    training_config = dict(
        optimizer='sgd',
        loss='nllloss'
    )
    trainer = TorchTrainer(
        model=Net(),
        config=training_config,
        epochs=2,
        strategy=strategy_name,
        checkpoint_every=1
    )
   
    strategy_instance = request.getfixturevalue(strategy_fixture)
    trainer.strategy = strategy_instance  # Patch the strategy with the fixture instance

    train_set, val_set = mnist_datasets
    trainer.execute(train_set, val_set)
``` 
I suppose it's debatable which one is more readable, and I'm not 100% it would work, but something to consider.

Reply:
interesting... I am using different markers for each test function in this case. Example:

```python
@pytest.mark.hpc
@pytest.mark.torch_dist
def test_distributed_trainer_ddp_mnist(mnist_datasets, ddp_strategy):
  ...

@pytest.mark.hpc
@pytest.mark.deepspeed_dist
def test_distributed_trainer_deepspeed_mnist(mnist_datasets, deepspeed_strategy):
  ...

@pytest.mark.hpc
@pytest.mark.horovod_dist
def test_distributed_trainer_horovod_mnist(mnist_datasets, horovod_strategy):
  ...
```

Do you know if specific markers could be added to the parametrization process somehow?